### PR TITLE
Additional changes to context pattern

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Removes value non-null validation, allowing for null values to be set and represent this value has been "removed".

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java [hunk 1]
```diff
@@ -71,12 +74,12 @@ public InternalContext put(Object key, Object value) {
...
     @Override
     public Object get(Object key) {
-        Object data = null;
+        Object data = SENTINEL;
 
         // Iterate in reverse order to get the most recent data first.
         for (int i = contexts.length - 1; i >= 0; i--) {
             data = contexts[i].get(key);
-            if (data != null) {
+            if (!Objects.equals(SENTINEL, data)) {
```
Comment (srnagar) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java:82]: Wouldn't `Objects.equals(SENTINEL, null)` return false?

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java [hunk 2]
```diff
@@ -71,12 +74,12 @@ public InternalContext put(Object key, Object value) {
 
     @Override
     public Object get(Object key) {
-        Object data = null;
+        Object data = SENTINEL;
 
         // Iterate in reverse order to get the most recent data first.
         for (int i = contexts.length - 1; i >= 0; i--) {
             data = contexts[i].get(key);
```
Comment (srnagar) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java:81]: Should we consider having `containsKey` method in InternalContext?

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java:82]: Correct, this is to allow `null` values to be set in Context and returned properly. If we had

- key: foo
\ - key2: bar
\ - key: null

We'd want null to be returned as it is last in the chain

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java:81]: We could, or we could have InternalContext use Optional whereas the public Context doesn't. I chose to go with the sentinel value as it is the cheapest computation wise.

Comment (JonathanGiles) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContextN.java:81]: I'm ok with this.

**Review by JonathanGiles — APPROVED** [review]

**Review by srnagar — APPROVED** [review]

**Pull request merged**